renderer/BloomStage.cpp

```diff
@@ -0,0 +1,321 @@
+	if (downsampleFBOs[0] != 0) {
+		qglDeleteFramebuffers(NUM_FILTER_STEPS, downsampleFBOs);
+		downsampleFBOs[0] = 0;
+	}
+	if (upsampleFBOs[0] != 0) {
+		qglDeleteFramebuffers(NUM_FILTER_STEPS, upsampleFBOs);
+		upsampleFBOs[0] = 0;
+	}
+	if (bloomDownSamplers != nullptr) {
+		bloomDownSamplers->PurgeImage();
+		bloomDownSamplers = nullptr;
+	}
+	if (bloomUpSamplers != nullptr) {
+		bloomUpSamplers->PurgeImage();
+		bloomUpSamplers = nullptr;
+	}
+}
+
+extern GLuint fboPrimary;
+extern bool primaryOn;
+
+void BloomStage::ComputeBloomFromRenderImage() {
+	GL_PROFILE("BloomStage");
+
+	int expectedWidth = r_fboResolution.GetFloat() * glConfig.vidWidth / 2;
+	int expectedHeight = r_fboResolution.GetFloat() * glConfig.vidHeight / 2;
+
+	if (downsampleFBOs[0] != 0 && (bloomDownSamplers->uploadWidth != expectedWidth || bloomDownSamplers->uploadHeight != expectedHeight)) {
+		// resolution changed, need to recreate our resources
+		Shutdown();
+	}
+
+	if (downsampleFBOs[0] == 0) {
+		Init();
+	}
+
+	qglClearColor(0, 0, 0, 0);
+	Downsample();
+	Blur();
+	Blur();
+	Upsample();
+
+	qglViewport( 0, 0, globalImages->currentRenderImage->uploadWidth, globalImages->currentRenderImage->uploadHeight );
+	// FIXME: this is a bit hacky, needs better FBO control
+	qglBindFramebuffer(GL_FRAMEBUFFER, primaryOn ? fboPrimary : 0);
+}
+
+void BloomStage::BindBloomTexture() {
+	bloomUpSamplers->Bind();
+}
+
+void BloomStage::Downsample() {
+	GL_PROFILE( "BloomDownsampling" )
+
+	// execute initial downsampling and bright pass on render image
+	downsampleWithBrightPassShader->Activate();
+	BloomDownsampleUniforms *uniforms = downsampleWithBrightPassShader->GetUniformGroup<BloomDownsampleUniforms>();
+	qglBindFramebuffer(GL_FRAMEBUFFER, downsampleFBOs[0]);
+	qglViewport( 0, 0, bloomDownSamplers->uploadWidth, bloomDownSamplers->uploadHeight );
+	uniforms->sourceMipLevel.Set( 0 );
+	uniforms->brightnessThreshold.Set( r_bloom_threshold.GetFloat() );
+	uniforms->thresholdFalloff.Set( r_bloom_threshold_falloff.GetFloat() );
+	GL_SelectTexture( 0 );
+	globalImages->currentRenderImage->Bind();
+	qglClear(GL_COLOR_BUFFER_BIT);
+	RB_DrawFullScreenQuad();
+
+	// generate additional downsampled mip levels
+	bloomDownSamplers->Bind();
+	downsampleShader->Activate();
+	uniforms = downsampleShader->GetUniformGroup<BloomDownsampleUniforms>();
+	for (int i = 1; i < NUM_FILTER_STEPS; ++i) {
+		uniforms->sourceMipLevel.Set( i - 1 );
+		qglBindFramebuffer(GL_FRAMEBUFFER, downsampleFBOs[i]);
+		qglViewport( 0, 0, bloomDownSamplers->uploadWidth >> i, bloomDownSamplers->uploadHeight >> i );
+		qglClear(GL_COLOR_BUFFER_BIT);
+		RB_DrawFullScreenQuad();
+	}
+}
+
+void BloomStage::Blur() {
+	GL_PROFILE("BloomBlur")
+
+	int mipLevel = NUM_FILTER_STEPS - 1;
+	blurShader->Activate();
+
+	BloomBlurUniforms *uniforms = blurShader->GetUniformGroup<BloomBlurUniforms>();
+	uniforms->mipLevel.Set( mipLevel );
+
+	uniforms->axis.Set( 1, 0 );
+	qglBindFramebuffer(GL_FRAMEBUFFER, upsampleFBOs[mipLevel]);
```

Comment:
`downsampleFBOs` and `upsampleFBOs` are swapped here?...

Reply:
The approach is basically:

- downsample to lowest mip in the downsamplers
- apply Gaussian blur to that lowest mip. This is a two-phase blur (vert/horz) and needs one intermediary texture to store the results of the first blur pass. I used the upsampling texture at lowest mip for that, because it seemed easier than creating yet another separate texture and framebuffer. Final result is stored back in the lowest mip of the downsamplers
- upsample, first from downsamplers[lowest_mip] to upsamplers[lowest_mip-1], then from upsamplers[mip+1] to upsamplers[mip]